### ISSUE TITLE
Build (and publish) macosArm64 and linuxX64

### ISSUE THREAD

**Issue by smyrgeorge** [moreirasantos/pgkn#51]
Hello!

I was searching for a native kotlin driver and I came across to you repo.

I saw that in the `build.gradle.kts` that you have a comment:
```text
// TODO find out how to get mac to compile linux64
```

We could have two separate builds (that run on 2 separate github actions) and each one will produce a different artifact.

For instance the `linuxX64` build will produce (and publish) the `io.github.moreirasantos:pgkn-linuxX64:x.y.z` artifact.
Accordingly the `macosArm64` will publish the `io.github.moreirasantos:pgkn-macosArm64:x.y.z` artifact.

Then in the `dependencies` section of the client project we can have something like this:
```kotlin
kotlin {
    sourceSets {
        val commonMain by getting {
            dependencies {
                ....
                when (System.getProperty("os.name")) {
                    "Mac OS X" -> {
                        implementation("io.github.moreirasantos:pgkn-macosArm64:1.1.0")
                    }

                    "Linux" -> {
                        implementation("io.github.moreirasantos:pgkn-linuxX64:1.1.0")
                    }
                }
            }
        }
    }
    ....
}
```

If it seems right to you, I could give it a shot and then provide you the PR.
Let me know 😄

**Comment by miguelsantosmindera** [comment]
Hello @smyrgeorge ! Thanks for the feedback!

You're right that this is a requirement for (sane) development of this library :).

However, you don't need to change the dependency section. That will fetch the proper library depending on the target being built. In fact, I already have a branch with support for other targets. https://github.com/moreirasantos/pgkn/blob/feat/support-macos-linux-arm-x64/build.gradle.kts

Looking at the branch, you have:

```kotlin
    macosArm64 { libpq("libpqArm.def") }
    macosX64 { libpq("libpqX.def") }
    linuxArm64 { libpq("libpqArm.def") }
    linuxX64 { libpq("libpqlinuxX.def") }
    jvm()
```
And what you're proposing is something like:
```kotlin
    when (System.getProperty("os.name")) {
        "Mac OS X" -> macosX64 { libpq("libpqX.def") }
        "Linux" -> linuxX64 { libpq("libpqlinuxX.def") }
    }
    jvm()
```

The only issue with that branch is that it compiles to all targets, it's not configurable.

For know, I would welcome a PR (probably based on above branch) where it would be possible to:
1. By default, only compile for the machine you're in (meaning, differentiate between linuxX64 and linuxArm64, for example - though it's acceptable to just choose one if it isn't possible)
2. Have a toggle to build to all targets at the same time (or a specific target)

My long-term plan is to have prebuilts in this repository, which would make compilation possible (in macos) regardless of what the machine has installed.

**Comment by moreirasantos** [comment]
Hello @smyrgeorge ! Thanks for the feedback!

You're right that this is a requirement for (sane) development of this library :).

However, you don't need to change the dependency section. That will fetch the proper library depending on the target being built. In fact, I already have a branch with support for other targets. https://github.com/moreirasantos/pgkn/blob/feat/support-macos-linux-arm-x64/build.gradle.kts

Looking at the branch, you have:

```kotlin
    macosArm64 { libpq("libpqArm.def") }
    macosX64 { libpq("libpqX.def") }
    linuxArm64 { libpq("libpqArm.def") }
    linuxX64 { libpq("libpqlinuxX.def") }
    jvm()
```
And what you're proposing is something like:
```kotlin
    when (System.getProperty("os.name")) {
        "Mac OS X" -> macosX64 { libpq("libpqX.def") }
        "Linux" -> linuxX64 { libpq("libpqlinuxX.def") }
    }
    jvm()
```

The only issue with that branch is that it compiles to all targets, it's not configurable.

For now, I would welcome a PR (probably based on above branch) where it would be possible to:
1. By default, only compile for the machine you're in (meaning, differentiate between linuxX64 and linuxArm64, for example - though it's acceptable to just choose one if it isn't possible)
2. Have a toggle to build to all targets at the same time (or a specific target)

My long-term plan is to have prebuilts in this repository, which would make compilation possible (in macos) regardless of what the machine has installed. This would also allow any linux machine to compile to x64 or arm64 without having libpq installed.

**Comment by smyrgeorge** [comment]
Thanks for your reply.

The problem is that I have a client project and I'm trying to build it (in my mac, arm64) and fails to download the dependency.

The problem (if I understand correctly) that the build (in github actions) is running in a linuxX64 machine.
So it never tries to build for the macosArm64.

PS: I forgot to mention the problem in the first message 😅. I went directly to the solution

**Comment by moreirasantos** [comment]
That is part of the issue, yes. Two things:

1. Current Github Workflow config doesn't support publishing
2. There is no macosArm64 target in main

I'll take a look today and get back to you.

Keep in mind that, even though you don't have main alternatives in this space, this library and most server-side Kotlin/Native tools are very experimental.

Meanwhile, you could use a DockerFile build and test your application, see example: https://github.com/moreirasantos/simple-ktor-native-server/blob/main/Dockerfile

Not sure if it unblocks you.

**Comment by smyrgeorge** [comment]
> 1. Current Github Workflow config doesn't support publishing

How you managed to publish to maven central?
Recently I was trying to publish to maven central and at the end I uploaded the artifacts manually.
I wrote a small guide here: [https://stackoverflow.com/questions/77763594/cannot-publish-to-maven-central-keep-getting-401](https://stackoverflow.com/questions/77763594/cannot-publish-to-maven-central-keep-getting-401)

> 2. There is no macosArm64 target in main

There is a macosArm64 runner. Take a look here: [https://github.com/actions/runner-images/blob/main/images/macos/macos-13-arm64-Readme.md](https://github.com/actions/runner-images/blob/main/images/macos/macos-13-arm64-Readme.md)

> Keep in mind that, even though you don't have main alternatives in this space, this library and most server-side Kotlin/Native tools are very experimental.

Yeah I know that it's very experimental, although I think that this stack has a lot of potential, so that's why I'm interested to dive in a bit. And I think this project is very important.

**Comment by moreirasantos** [comment]
> How you managed to publish to maven central?

With the proper credentials and PGP private keys, simply run in this project:

```shell
./gradlew clean publishAllPublicationsToSonatypeRepository
```

Now there's a maven-publish gradle plugin which I haven't tried, but current implementation is done similar to this guide:

https://dev.to/kotlin/how-to-build-and-publish-a-kotlin-multiplatform-library-going-public-4a8k

> There is a macosArm64 runner

Yes, I mean that main branch isn't configured with macos target.

**Comment by moreirasantos** [comment]
pgkn version `1.2.0` released, with macosArm64 support:

https://central.sonatype.com/artifact/io.github.moreirasantos/pgkn
https://central.sonatype.com/artifact/io.github.moreirasantos/pgkn-macosarm64

Reminder that for your `macosArm64` target in your kotlin native project, this should work:

```
                implementation("io.github.moreirasantos:pgkn:1.2.0")
```
(No need to specify platform.)

**Comment by moreirasantos** [comment]
Closing this Issue, let me know if you need anything else.

I've also created another issue regarding publishing in CI/CD specifically: https://github.com/moreirasantos/pgkn/issues/53

Thank you again for your feedback!

**Comment by smyrgeorge** [comment]
Thanks a lot! Works perfectly now!
